platform/darwin/src/NSExpression+MGLAdditions.mm

```diff
@@ -1100,4 +1100,199 @@ - (id)mgl_jsonHasExpressionObject {
+            }
+            localizedCollection[idx] = localizedItem;
+        }
+    }];
+    return localizedCollection ?: collection;
+};
+
+/**
+ Returns a localized copy of the given stop dictionary.
+ 
+ If no localization takes place, this method returns the original stop
+ dictionary.
+ */
+NS_DICTIONARY_OF(NSNumber *, NSExpression *) *MGLLocalizedStopDictionary(NS_DICTIONARY_OF(NSNumber *, NSExpression *) *stops, NSLocale * _Nullable locale, BOOL replacesTokens) {
+    __block NSMutableDictionary *localizedStops;
+    [stops enumerateKeysAndObjectsUsingBlock:^(NSNumber * _Nonnull zoomLevel, NSExpression * _Nonnull value, BOOL * _Nonnull stop) {
+        NSExpression *localizedValue = [value mgl_expressionLocalizedIntoLocale:locale replacingTokens:replacesTokens];
+        if (localizedValue != value) {
+            if (!localizedStops) {
+                localizedStops = [stops mutableCopy];
+            }
+            localizedStops[zoomLevel] = localizedValue;
+        }
+    }];
+    return localizedStops ?: stops;
+};
+
+- (NSExpression *)mgl_expressionLocalizedIntoLocale:(nullable NSLocale *)locale replacingTokens:(BOOL)replacesTokens {
+    NSString *localizedKeyPath = @"name";
+    if (locale && ![locale.localeIdentifier isEqualToString:@"mul"]) {
+        localizedKeyPath = [NSString stringWithFormat:@"name_%@", locale.localeIdentifier];
+    }
+    switch (self.expressionType) {
+        case NSConstantValueExpressionType: {
+            if (replacesTokens) {
+                NSString *constantValue = self.constantValue;
+                if ([constantValue isKindOfClass:[NSString class]]
+                    && ([constantValue containsString:@"{name}"] || [constantValue containsString:@"{name_"])) {
+                    return [NSExpression expressionForKeyPath:localizedKeyPath];
+                }
+            }
+            NSDictionary *stops = self.constantValue;
+            // TODO: Check whether the dictionary’s key and value types are consistent with stop dictionaries. Or have the caller pass in whether this is a stop dictionary.
+            if ([stops isKindOfClass:[NSDictionary class]]) {
+                NSDictionary *localizedStops = MGLLocalizedStopDictionary(stops, locale, replacesTokens);
+                if (localizedStops != stops) {
+                    return [NSExpression expressionForConstantValue:localizedStops];
+                }
+            }
+            return self;
+        }
+            
+        case NSKeyPathExpressionType: {
+            if ([self.keyPath isEqualToString:@"name"] || [self.keyPath hasPrefix:@"name_"]) {
+                return [NSExpression expressionForKeyPath:localizedKeyPath];
+            }
+            return self;
+        }
+            
+        case NSFunctionExpressionType: {
+            NSExpression *operand = self.operand;
+            NSExpression *localizedOperand = [operand mgl_expressionLocalizedIntoLocale:locale replacingTokens:NO];
```

Comment:
According to this:
> @param replacesTokens `YES` to localize `{token}` syntax in constant value subexpressions; `NO` otherwise.

`NSExpression *localizedOperand = [operand mgl_expressionLocalizedIntoLocale:locale replacingTokens:NO];` overrides the value passed in `replacesTokens`. What would happen if the function expression contains a constant value? should not this be behave accordingly?

Reply:
Ah, this deserves clarification: the `replacesTokens` parameter is only meant to replace the legacy token syntax, but that’s a subset of the syntax that we want to localize. The only reason that the `replacesTokens` functionality is supported at all in functions is that three functions in particular may represent style functions, which can contain tokens. Nothing else in the expression language can contain tokens.

Having said all that, #11659 means that tokens are only supported for the `MGLSymbolStyleLayer.text` and `iconImageName` properties’ getters, not their setters. So as long as we come up with an expression to set, we need to upgrade all the tokens within the expression to key path expressions. 023624360aa37c4d984b3f49484c8d3552b2b414 does that, but only as a post-localization step on individual function calls.

We may end up needing to unconditionally upgrade the entire expression upfront, before even looking at whether to localize anything. If we do that, then the separate `replacesTokens` functionality will no longer be necessary.